aea/aea_builder.py

```diff
@@ -0,0 +1,736 @@
+
+        component_ids = itertools.chain(
+            [
+                ComponentId(ComponentType.PROTOCOL, p_id)
+                for p_id in agent_configuration.protocols
+            ],
+            [
+                ComponentId(ComponentType.CONNECTION, p_id)
+                for p_id in agent_configuration.connections
+            ],
+            [
+                ComponentId(ComponentType.SKILL, p_id)
+                for p_id in agent_configuration.skills
+            ],
+        )
+        for component_id in component_ids:
+            component_path = cls._find_component_directory_from_component_id(
+                aea_project_path, component_id
+            )
+            builder.add_component(component_id.component_type, component_path)
+
+        return builder
+
+
+def _verify_or_create_private_keys(aea_project_path: Path) -> None:
+    """Verify or create private keys."""
+    path_to_configuration = aea_project_path / DEFAULT_AEA_CONFIG_FILE
+    agent_loader = ConfigLoader("aea-config_schema.json", AgentConfig)
+    fp_read = path_to_configuration.open(mode="r", encoding="utf-8")
+    agent_configuration = agent_loader.load(fp_read)
+
+    for identifier, _value in agent_configuration.private_key_paths.read_all():
+        if identifier not in SUPPORTED_CRYPTOS:
+            ValueError("Unsupported identifier in private key paths.")
+
+    fetchai_private_key_path = agent_configuration.private_key_paths.read(FETCHAI)
+    if fetchai_private_key_path is None:
+        _create_fetchai_private_key(
+            private_key_file=str(aea_project_path / FETCHAI_PRIVATE_KEY_FILE)
+        )
+        agent_configuration.private_key_paths.update(FETCHAI, FETCHAI_PRIVATE_KEY_FILE)
+    else:
+        try:
+            _try_validate_fet_private_key_path(
+                str(aea_project_path / fetchai_private_key_path), exit_on_error=False
+            )
+        except FileNotFoundError:  # pragma: no cover
+            logger.error(
+                "File {} for private key {} not found.".format(
+                    repr(fetchai_private_key_path), FETCHAI,
+                )
+            )
+            raise
+
+    ethereum_private_key_path = agent_configuration.private_key_paths.read(ETHEREUM)
+    if ethereum_private_key_path is None:
+        _create_ethereum_private_key(
+            private_key_file=str(aea_project_path / ETHEREUM_PRIVATE_KEY_FILE)
+        )
+        agent_configuration.private_key_paths.update(
+            ETHEREUM, ETHEREUM_PRIVATE_KEY_FILE
+        )
+    else:
+        try:
+            _try_validate_ethereum_private_key_path(
+                str(aea_project_path / ethereum_private_key_path), exit_on_error=False
+            )
+        except FileNotFoundError:  # pragma: no cover
+            logger.error(
+                "File {} for private key {} not found.".format(
+                    repr(ethereum_private_key_path), ETHEREUM,
+                )
+            )
+            raise
```

Comment:
This function is copied from aea.cli.common._verify_or_create_private_keys, with some variation. It is used in AEABuilder.from_aea_config in order to emulate the behaviour of aea run. In particular, the CLI command instantiates the private key if it is not in the project path. Of course it can be further improved and refactored.